### ISSUE TITLE
E2E tests: move the "test controller" to a container within the Kubernetes cluster itself

### ISSUE THREAD

**Issue by ItalyPaleAle** [dapr/dapr#4447]
## In what area(s)?

/area test-and-release

## Describe the proposal

Currently, E2E tests are designed to be executed by a "test controller" app that runs "locally" (in the case of the GitHub Action, that means within the Actions runner itself). This "test controller" communicates with the K8s cluster to create new apps, then connects to the apps to verify their behavior.

When the tests run on a K8s cluster that's not in the runner itself, the communication goes over some network.

During my recent work on improving E2E tests, I noticed that the network between the Actions runner and AKS is not always reliable. Tests are not-infrequently failing due to network issues (connections dropped, timeouts, etc).

In some cases these are not transient: even when a retry logic is in place, the connection kept failing for at least 10 minutes. It's unclear exactly what causes this, but tests have been showing this behavior pretty much consistently when targeting AKS in Canada Central or East Asia.

Even when the failures are transient and a retry would fix them, it would still require a lot of work to go through our test infrastructure and add some resiliency support (in some cases operations aren't idempotent either, so automatically re-trying them in case of network failures may not be ideal).

A possible solution to this problem could be to containerize the "test controller" and run it inside the K8s cluster itself. So all communication would be "local" inside the K8s cluster.

Inside Actions, the Runner would create a container of the test app and deploy that like any other workload on K8s. The test controller would do what it needs to do and then return the response to the Actions Runner. This way I hope we would eliminate flakiness due to issues with the network.

(Note that this does **not** require the creation of self-hosted Actions runner - we'd still be using the public Actions Runners, but their job would more limited to just kickstarting the tests and waiting for the results).

**Comment by artursouza** [comment]
I am supportive of running the test code from within the AKS cluster itself. My only concern would be if that can be achieved without moving the GitHub Actions Runner entirely too.

**Comment by ItalyPaleAle** [comment]
Yes, not moving the Actions Runner (that is, not needing to create self-hosted ones) is a requirement.

**Comment by ItalyPaleAle** [comment]
My hope is that given that this would make less network calls and less frequently, it should be more stable.

If the runner couldn't even make a single call to deploy a pod, and then another one to retrieve the results, that would be really bad

**Comment by ItalyPaleAle** [comment]
After the changes made in #4626 we have seen stabilization in the E2E tests. Network timeouts seemed to have been primarily caused by SNAT port exhaustion in the Actions runners. Mitigations implemented in #4626 seem to have significantly reduced those kinds of issues (although there are other issues to fix :) ).

I am not sure if this work is still needed now.
